keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,854 @@
+of the Job created by the user.
+
+In order to support this workflow we need a mechanism to disable the main Job
+controller, and delegate the status synchronization to the Kueue controller.
+
+### Goals
+
+- support delegation of Job synchronization to an external controller
+
+### Non-Goals
+
+- passing custom parameters to the external controller
+
+## Proposal
+
+The proposal is to support the `managed-by` label to disabled the Job controller
+from synchronizing jobs with the annotation.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a developer of Kueue I want to have Job API which allows me to implement the
+MultiKueue design. For this reason I need a way to disable the main Job
+controller on the management cluster.
+
+The mechanism should be per-Job, because the management cluster might also be
+one of the worker clusters, for two reasons:
+1. Disabling the Job controller per cluster requires access the `kube-controller-manager`
+   manifest. Such access is generally discouraged by cloud providers.
+2. The management cluster may also be a worker. Supporting this scenario is important
+   for smooth transition of Kueue users from a single-cluster to multi-cluster.
+
+Ideally, I prefer the mechanism to be lightweight so that it is easy to be adopted
+by other Job CRDs supported by Kueue
+(see [here](https://github.com/kubernetes-sigs/kueue/blob/6d428f3279a9ca0e204c083dc649dbbc6558db71/config/components/manager/controller_manager_config.yaml#L31-L42)):
+MPIJob, RayJob, JobSet, multiple Kubeflow jobs.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Prior work
+
+This approach of allowing another controller to mirror information between APIs
+is already supported with the `managed-by` label used by
+EndpointSlices (`endpointslice.kubernetes.io/managed-by`) and IPAddresses
+(`ipaddress.kubernetes.io/managed-by`).
+
+#### Graduating directly to Beta
+
+The implementation is simple (see [Design details](#design-details)), short-circuit the
+`syncJob` invocation. Also, since the feature is just label-based, there is
+technically no need to go via Alpha in order to ensure that all
+kube-apiserver instances (in HA setup, see [here](https://kubernetes.io/releases/version-skew-policy/#supported-version-skew))
+recognize the new API, as in case of new fields. A similar approach was used
+to graduate the
+[Pod Index Label for StatefulSets and Indexed Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4017-pod-index-label)
+feature with similar characteristics.
+
+Finally, we would like to have the feature soak without unnecessary delay.
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
```

Comment:
What happens if you add `managed-by` for a controller that doesn't exist, would that be equivalent to setting `suspect=true` and it would hang around until/if the controller exists? Or to flip that scenario, what if the controller does exist but then goes away (what happens to the jobs)? This is the ephemeral worker cluster case.

Reply:
> What happens if you add managed-by for a controller that doesn't exist, 

Then it is not synchronized by any controller, the status is left stale.

> would that be equivalent to setting suspect=true and it would hang around until/if the controller exists? 

I assume you ask `suspend=true`, no this is not equivalent. See also in alternatives: https://github.com/kubernetes/enhancements/blob/340c1d3d969826baa1dd31e6f6656396b3e76c52/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md#use-the-specsuspend-field

> Or to flip that scenario, what if the controller does exist but then goes away (what happens to the jobs)?

From the core k8s perspective no difference, the built-in Job controller will skip, what happens with the other controller is out of the control. It is up to the platform administrators to ensure the system is configured properly as a whole.